p4-16/spec/P4-16-spec.mdk

```diff
@@ -4762,6 +4768,36 @@ H h2 = { f2=5, ... };   // initialize h2 with a header that is valid, field f1 0
 H h3 = { ... };  // initialize h3 with a header that is valid, field f1 0, field f2 0
 ~ End P4Example
 
+# Compile-time size determination { #sec-minsizeinbits }
+
+The method calls `minSizeInBits` and `minSizeInBytes` can be applied
+to most expressions.  Each of these method calls evaluate to
+compile-time known values that return the minimum size in bits
+required to store the expression.
+
+The method `minSizeInBytes` always returns the result the result of
+`minSizeInBits` rounded up to the next whole number of bytes.  not a
+multiple of 8 bits long.  In other words, for any expression `e`,
+`e.minSizeInBytes()` is equal to `(e.minSizeInBits() + 7) >> 3`.
+
+The definition of `e.minSizeInBits` is given recursively on the type of `e` as
+follows:
+
+* if the type of `e` is `varbits` the result is 0
+* if the type of `e` is `bit<N>` or `int<N>` the result is `N`
+* if the type of `e` is a `struct` or `header`, the result is the sum
+  of the `minSizeInBits` of all fields of the struct or header.
+* if the type of `e` is a `header_union` the result is the max of the
+  `minSizeInBits` of all fields of the union
+* if the type of `e` is a header stack, the result is the size of the
+  stack multiplied with the `minSizeInBits` of an element
+* if the type of `e` is a serializable `enum`, the result is the
+  `minSizeInBits` of the underlying type
+* if the type of `e` is a type introduced by `newtype`, the result
```

Comment:
Do we want to enforce stricter information hiding for `newtype`?

Reply:
@jnfoster -- I do not think we need to hide the size of the `newtype`, becaue it still ahs to be visible via table keys, action parameters, etc. The only thing we really want to do (in that specific case where the type is used for port (re)mapping is to constrain some operations that might be dangerous (e.g. ternary matching, math, etc.). Note by the way that today the constraints are so severe that this construct is almost useless.